### ISSUE TITLE
Feature Request: Detatch Instance from ASG upon Termination Event

### ISSUE THREAD

**Issue by jtcressy** [aws/aws-node-termination-handler#141]
The other termination-handler project at https://github.com/kube-aws/kube-spot-termination-notice-handler is capable of detaching the instance from its ASG when a termination event occurs. [Link](https://github.com/kube-aws/kube-spot-termination-notice-handler#autoscaling-detachment) 

This feature is useful to compensate for the extremely long lag time between instance termination and another ASG replacement coming online. If the instance is detached from the asg immediately, a new node can be starting up during the 120s grace period.

Currently, my only method of ensuring minimum compute capacity is to run an extra node or two at all times so that pods get rescheduled onto other nodes when one of them is drained. If a new instance comes online fast enough for the drained pods to reschedule, the problem is solved.

TL;DR: Remove instance from ASG early to bring new nodes online faster.

**Comment by bwagner5** [comment]
Thanks for your interest in Node Termination Handler, and apologies for the delay in responding!

First, the bad news. The current design of Node Termination Handler does not allow us to manipulate AWS resources like ASGs. In order for NTH to do that, we would need a more complicated setup process to get credentials. We've discussed at length in this issue https://github.com/aws/aws-node-termination-handler/issues/14 about why we chose to do this. Basically, we would need to create a kubernetes controller or operator rather than a daemonset, which is what we have now.

We are exploring an implementation that would have credentials that could do things like detach from a spot instance from an ASG when an interruption notice is received. This will take some time to think through how this would work and how to integrate it with the current node termination handler though. 

We will keep this issue updated with any progress on that front and would welcome any feedback on the direction we go!

**Comment by kzap** [comment]
Running cluster autoscaler would compensate for this and bring up new nodes for you when needed rather than waiting for one to be replaced?

**Comment by bwagner5** [comment]
If a spot node was interrupted and aws-node-termination-handler cordoned and drained the node, then pods would go into pending and CA should scale up. I think the optimization that this issue is tracking is the amount of time it takes to get another node up. There could be latency in CA scaling up, but if we detach the node from the ASG, then the ASG would strive for the desired node count and scale up automatically.  This case is probably less likely to matter. 

I think a better scenario is when the ASG is at max capacity and CA is not able to scale up because of that. It would then have to wait until the interruption termination notice window is up and then scale up, which would cause unnecessary latency in scaling. Since aws-node-termination-handle knows that a node is about to disappear due to a spot interruption, detaching from the ASG would allow for an additional node to be brought up without waiting.

**Comment by cshivashankar** [comment]
Hi @bwagner5  This is the only feature which is stopping us from moving to aws termination handler. At present custom solution is being implemented.But I would love to use the amazon version.
Any idea when this feature  could be released.

**Comment by gabegorelick** [comment]
> First, the bad news. The current design of Node Termination Handler does not allow us to manipulate AWS resources like ASGs. In order for NTH to do that, we would need a more complicated setup process to get credentials. We've discussed at length in this issue #14 about why we chose to do this. Basically, we would need to create a kubernetes controller or operator rather than a daemonset, which is what we have now.

Is this still applicable now that queue processor mode is a thing?

**Comment by bwagner5** [comment]
It is now possible natively within EC2 AutoScaling Groups to proactively launch a new instance when a spot instance receives an interruption termination notification (ITN). 

You can enable `--capacity-rebalance` on the ASG which tells the ASG to monitor for rebalance events (which could occur before an ITN) so that you can proactively replace spot instances that may be interrupted soon, and have more time to do so than the 2 minutes w/ ITN.  When ASG receives a rebalance event, it will first launch the replacement instance, and once online, it will terminate the spot instance that received the rebalance. 

Additionally, if you enable `--capacity-rebalance` ASG will also monitor for Spot ITNs. When it receives an ITN, ASG will launch a new instance at the same time that it terminates the spot instance. 

[ASG Capacity Rebalance docs](https://docs.aws.amazon.com/autoscaling/ec2/userguide/capacity-rebalance.html)
[Capacity Rebalance Blog Post](https://aws.amazon.com/blogs/compute/proactively-manage-spot-instance-lifecycle-using-the-new-capacity-rebalancing-feature-for-ec2-auto-scaling/) 

I believe this feature will take care of your request without any logic being added to aws-node-termination-handler.

**Comment by bwagner5** [comment]
Feel free to reopen if you think this is still an issue.

**Comment by jmriebold** [comment]
Hey @bwagner5, we've been using kube-spot-termination-notice-handler for quite a while, but recently decided to try out node-termination-handler now that the capacity rebalance feature exists. Unfortunately it didn't work out for us at all.

We enabled the capacity rebalance feature on our ASGs, then deployed NTH with the following parameters:

```yaml
enableSpotInterruptionDraining: true
enableScheduledEventDraining: true
enableRebalanceMonitoring: true
enableRebalanceDraining: true
```

however the result was all our spot nodes endless thrashing as they'd spin up and then seemingly immediately receive a rebalance notification (often <5 min), cordon and drain, spin up a replacement, rinse and repeat over and over and over. I don't fully understand how rebalance notifications work, but it seems to me that the notifications can come _far_ in advance of the actual spot termination notice, which dramatically shortens the average lifespan of a spot node.

On the other hand, if `enableRebalanceMonitoring` isn't enabled, we'll lose nodes without draining them at all (because the ASG terminates the existing node once its replacement is healthy). But if we disable capacity rebalance we're right back at square one, namely that 2 minutes is not enough time to both trigger a cluster scale-up _and_ drain the terminating node, resulting in frequent short outages every time a spot node is terminated and pods get rescheduled. KSTNH's approach of detaching the terminating node from the ASG results in the replacement node being added just quickly enough that we're usually able to fully drain the node before termination, so that's what we're sticking with for now.

**Comment by bwagner5** [comment]
It looks like you're using IMDS mode which doesn't directly integrate with ASGs so there can be some fighting between the two systems.  NTH Queue-Processor mode will be a better experience in general and handles Rebalance notifications at the ASG level.  So, on a rebalance event, the ASG will spin up a replacement instance before it starts draining the instance that received the rebalance recommendation. Only after the new instance is ready to go will the ASG start a termination lifecycle hook which NTH queue-processor mode will receive and then start the cordon and drain workflow. I think this is the behavior you are after when wanting the detachment from the ASG?  

I'm not sure why you are seeing so many rebalance recommendations. Are you supplying a few different instance types for the ASG to choose from? If the ASG is just configured with one instance type, you might receive more rebalance recommendations.

**Comment by jmriebold** [comment]
Yeah, sorry, should've mentioned I was running in IMDS mode.

> Only after the new instance is ready to go will the ASG start a termination lifecycle hook which NTH queue-processor mode will receive and then start the cordon and drain workflow.

Yeah this is exactly what I'm after. I think it would actually be an improvement on what we've got going now with KSTNH since it should guarantee a full 2 minutes instead of a bit less than that.

> I'm not sure why you are seeing so many rebalance recommendations.

I'm not sure either, and now I wish I would've grabbed some of the logs from the NTH pods. Maybe I'll redeploy NTH and see if I can figure out what's going on.

> Are you supplying a few different instance types for the ASG to choose from? If the ASG is just configured with one instance type, you might receive more rebalance recommendations.

Our ASGs have two instance types each (e.g. m5.2xlarge, m5.4xlarge).

Anyways, I'm gonna try redeploying NTH in IMDS mode to grab some logs, then I'll try queue processor mode afterwards. I'll let you know what I find out.

**Comment by bwagner5** [comment]
Nice!  You might want to add some more instance types to those ASGs to reduce interruptions and rebalance events. My team wrote a tool that helps find instance types that fulfill a certain resource criteria: https://github.com/aws/amazon-ec2-instance-selector

You can do things like:
```
$ ec2-instance-selector --vcpus=8 --memory-min=16gb --gpus=0 --cpu-architecture amd64 -o table-wide
Instance Type  VCPUs   Mem (GiB)  Hypervisor  Current Gen  Hibernation Support  CPU Arch  Network Performance  ENIs    GPUs    GPU Mem (GiB)  GPU Info
-------------  -----   ---------  ----------  -----------  -------------------  --------  -------------------  ----    ----    -------------  --------
c5.2xlarge     8       16.000     nitro       true         true                 x86_64    Up to 10 Gigabit     4       0       0.00
c5a.2xlarge    8       16.000     nitro       true         false                x86_64    Up to 10 Gigabit     4       0       0.00
c5ad.2xlarge   8       16.000     nitro       true         false                x86_64    Up to 10 Gigabit     4       0       0.00
c5d.2xlarge    8       16.000     nitro       true         false                x86_64    Up to 10 Gigabit     4       0       0.00
c5n.2xlarge    8       21.000     nitro       true         false                x86_64    Up to 25 Gigabit     4       0       0.00
d2.2xlarge     8       61.000     xen         true         false                x86_64    High                 4       0       0.00
d3.2xlarge     8       64.000     nitro       true         false                x86_64    Up to 15 Gigabit     4       0       0.00
d3en.2xlarge   8       32.000     nitro       true         false                x86_64    Up to 25 Gigabit     4       0       0.00
f1.2xlarge     8       122.000    xen         true         false                x86_64    Up to 10 Gigabit     4       0       0.00
h1.2xlarge     8       32.000     xen         true         false                x86_64    Up to 10 Gigabit     4       0       0.00
i2.2xlarge     8       61.000     xen         false        false                x86_64    High                 4       0       0.00
i3.2xlarge     8       61.000     xen         true         true                 x86_64    Up to 10 Gigabit     4       0       0.00
i3en.2xlarge   8       64.000     nitro       true         false                x86_64    Up to 25 Gigabit     4       0       0.00
inf1.2xlarge   8       16.000     nitro       true         false                x86_64    Up to 25 Gigabit     4       0       0.00
m2.4xlarge     8       68.399     xen         false        false                x86_64    High                 8       0       0.00
m3.2xlarge     8       30.000     xen         false        true                 x86_64    High                 4       0       0.00
m4.2xlarge     8       32.000     xen         true         true                 x86_64    High                 4       0       0.00
m5.2xlarge     8       32.000     nitro       true         true                 x86_64    Up to 10 Gigabit     4       0       0.00
m5a.2xlarge    8       32.000     nitro       true         true                 x86_64    Up to 10 Gigabit     4       0       0.00
m5ad.2xlarge   8       32.000     nitro       true         true                 x86_64    Up to 10 Gigabit     4       0       0.00
NOTE: 18 entries were truncated, increase --max-results to see more
```

**Comment by jmriebold** [comment]
Very cool, I'll check that out. Thanks for the recommendation.

With regard to interruptions though, we typically only see 5-20 per day for our dev cluster (100% spot, 5-10 nodes), which isn't too disruptive. Some nodes will go a day or two before getting terminated.

**Comment by jmriebold** [comment]
Ok here's a log. It looks fine so far as I can tell, apart from the fact that it got a rebalance notification almost immediately after starting.

For reference, here are the values we're using with the Helm chart:
```yaml
awsRegion: "us-west-2"
enableSpotInterruptionDraining: true
enableScheduledEventDraining: true
enableRebalanceMonitoring: true
enableRebalanceDraining: true
jsonLogging: true
deleteLocalData: true
ignoreDaemonSets: true
podAnnotations:
  iam.amazonaws.com/role: xxxxx
webhookURL: https://hooks.slack.com/services/xxxxx
webhookTemplate: '{"attachments": [{"fallback": "Spot Interruption: {{ .Kind }}, Node: {{ .NodeName }}, Type: {{ .InstanceType }}, Action: {{ .Description }}", "title": ":warning: Spot Interruption", "color": "${color}", "fields": [{"title": "Type", "value": "{{ .Kind }}", "short": false}, {"title": "Node", "value": "{{ .NodeName }}", "short": false}, {"title": "Instance Type", "value": "{{ .InstanceType }}", "short": false}, {"title": "Action", "value": "{{ .Description }}", "short": false}]}]}'
enablePrometheusServer: true
enableProbesServer: true
podMonitor:
  create: true
tolerations:
  - ...
useHostNetwork: false
```

[nth.log](https://github.com/aws/aws-node-termination-handler/files/6401682/nth.log)

**Comment by jmriebold** [comment]
Hey @bwagner5, just went through the setup for deploying NTH in queue-processor mode. My plan is to monitor the queue for a bit to compare the rebalance events there vs. what KSTNH is doing before deploying NTH again. One question though: should I enable the capacity rebalance feature enabled for the ASGs? Or does queue-processor mode's use of the lifecycle hooks mean that's not needed?

**Comment by jmriebold** [comment]
Alright, I left the EventBridge rules enabled overnight (approximately 17 hours) without any consumers monitoring the queue, and dumped the resulting 45 messages this morning. Of the 45, 30 are rebalance notifications, and only 15 are spot terminations. Grouping by instance id we get the following:

**Nodes with rebalance notification and termination notice**: 9
**Nodes with rebalance notification and no termination notice**: 21
**Nodes with termination notice and no rebalance notification**: 6

For those nodes that received a rebalance notification and were subsequently terminated, the interval between rebalance notification and termination notice ranges from 17 minutes to 8 hours and 42 minutes, which is a massive range.

The question is why are we getting so many rebalance notifications? I spent some reading the documentation around them, and I think I can answer that too. My hypothesis is that our ASGs are becoming unbalanced across AZs (each deploys to 3 AZs), and that the rebalance notifications are being sent due to that. ASGs becoming unbalanced across AZs when used with a Kubernetes cluster is virtually guaranteed if the cluster scales in and out enough, since Cluster Autoscaler selects nodes for termination based on usage, not AZs. This is an issue we became aware of quite a while ago, and to work around it we disabled the AZRebalance process on our ASGs., but it seems the rebalance notifications are ignoring that.

So to sum up: we can't use NTH with rebalance notifications because they're being sent for reasons other than impending spot termination in a way that's very disruptive. But we also can't use NTH with termination notices only because we could end up not having enough time to drain a node since NTH doesn't detach terminating nodes from the ASG to prompt the creation of a replacement node like KSTNH does.

**Comment by bwagner5** [comment]
Rebalance Recommendation Events only occur due to impending Spot Interruption Notices, AZ rebalance events cannot cause a rebalance recommendations (the naming is slightly unfortunate).  So even if you ignore rebalance events, you will eventually get a spot interruption notice, it just depends on how long between the rebalance recommendation and the spot interruption notice.  

I'm curious if the ones you didn't receive a spot interruption notice on but did a rebalance event were close to the time you dumped the queue messages?  I would suspect those spot interruption notices would be coming in within 10-20 minutes. 

We are aware of the cluster autoscaler issues with AZ rebalance and generally the guidance is to disable, as you have done, and use a separate ASG per AZ.

**Comment by jmriebold** [comment]
Interesting. In that case I truly don't understand what's going on with these rebalance notifications. I just grabbed the terminations that have occurred between when I consumed the queue and now (about 4.5 hours have elapsed), and the totals now stand as follows:

**Nodes with rebalance notification and termination notice**: 14
**Nodes with rebalance notification and no termination notice**: 16

As for when the rebalance notifications arrive, the majority seem to arrive well before 10-20 minutes before termination. Here are the intervals for the instances that have received both a rebalance notice and a termination notice:

```
0 hours 16 minutes
0 hours 18 minutes
1 hours 40 minutes
1 hours 45 minutes
1 hours 46 minutes
4 hours 15 minutes
4 hours 53 minutes
5 hours 47 minutes
6 hours 17 minutes
7 hours 9 minutes
7 hours 10 minutes
7 hours 20 minutes
8 hours 3 minutes
8 hours 42 minutes
```

As for the remaining 16 nodes which have received a rebalance notice, but not yet received a termination notice, here's how much time has elapsed since the rebalance notice (calculated as of 1:45 PM PDT):

```
4 hours 3 minutes
4 hours 15 minutes
4 hours 16 minutes
4 hours 23 minutes
4 hours 23 minutes
4 hours 23 minutes
4 hours 25 minutes
4 hours 37 minutes
5 hours 16 minutes
5 hours 37 minutes
6 hours 18 minutes
6 hours 22 minutes
8 hours 40 minutes
11 hours 41 minutes
20 hours 38 minutes
21 hours 1 minutes
```

So we're approaching a full 24 hours for a couple of these instances...

**Comment by jmriebold** [comment]
Will do. Definitely looks like unexpected behavior, but I can't imagine what else would be causing it.

**Comment by jmriebold** [comment]
Hey @bwagner5, heard back from AWS over the weekend, and it seems like this is normal behavior. The CSR said the following:

> As can be seen from [your] examples, there may only be a few minutes, but there could also be a few hours between the two events. As per our documentation [1], the event that indicates the "EC2 Instance Rebalance Recommendation" is merely a signal that indicates that a Spot Instance "is at elevated risk of interruption" - it does not indicate that an instance will be interrupted, but the "EC2 Spot Instance Interruption Warning" is an indication that the Spot Instance will be terminated within the next two minutes.

> So while the 1st event "EC2 Instance Rebalance Recommendation" may be used to take some proactive actions such as gracefully shutting down a spot instance, preventing new work from being scheduled on that instance or even launch replacement instances, it should never be interpreted as a signal that an instance will be terminated soon - it merely indicates that there is a risk that an instance may be terminated soon, but the "EC2 Spot Instance Interruption Warning" is a confirmation that an instance will be terminated soon.

**Comment by stevehipwell** [comment]
@bwagner5 what would you expect to happen when a rebalance notice is evaluated? We're seeing NTH draining the node but then cluster-autoscaler increases the desired capacity based on the extra node running that then throws the autoscaler behaviour off.

As a thought would it be possible to flag the node to be cleaned up by auto-scaler after it's drained so it doesn't add to the desired capacity?

**Comment by bwagner5** [comment]
Is the ASG that CA is managing configured with capacity-rebalanced enabled?  If that’s enabled, then ASG should terminate the old instance when the new one has come online.

**Comment by stevehipwell** [comment]
@bwagner5 it is configured in the ASG. Basically before the ASG can terminate the node CA picks it up and modifies the desired capacity for the ASG which then results in a long running reconciliation loop for CA resulting in a not insignificant node churn. If there are labels or taints that make the node either invisible or deleteable to CA they would be worth setting. As it is I've had to disable capacity rebalance until a solution can be found.

**Comment by hamishforbes** [comment]
I've been having the same issues as everyone else here
Rebalance recommendations are too frequent and inaccurate (too many unnecessary terminations)
Cluster Autoscaler +1s the desired capacity when a node gets drained, the node is then terminated and a new node replaces it, the CA then -1s the desired capacity as the cluster is now +1 compared to before the drain

Rather than try to shoehorn the ASG detach mechanism into the NTH i knocked up [aws-node-asg-detach](https://github.com/hamishforbes/aws-node-asg-detach).
It's not perfect because the terminating node is still unschedulable for a time before the replacement is ready, but it does significantly reduce the time for a replacement node to be available.

I only wrote it yesterday so still in testing, not running on production, use at your own risk etc etc!

**Comment by SCLogo** [comment]
do you still disable the rebalance recommendation or do you have any setup for CA to not get spot instances with rebalance recomm?

we are facing this issue very often... sometimes we have 20 more nodes before find one without rebalance signal. 
Thanks
